### ISSUE TITLE
Concurrency

### ISSUE THREAD

**Issue by stevencox** [databricks/spark-xml#130]
Can spark-xml be used to read lots of files in parallel? i.e., I've used RDDs of file names in the past and, using map(), read the files in parallel at the executors. The general usage patterns I see for spark-xml look like I/O would be serialized at the driver. Am I missing something simple?

Thanks.

**Comment by HyukjinKwon** [comment]
Hi @stevencox, the transformation process would be executed in parallel just like the other data sources. Do you mind if I ask some sample codes you run? I would like to reproduce if it was observed that the jobs are executed not in parallel.

**Comment by stevencox** [comment]
Hi @HyukjinKwon, I'm probably just not using the API correctly. Perhaps you can help me get it straight.

I'm iterating over a list of XML files to create a list of dataframe objects. 

My first question is why there's a large pause between each file (which suggest it's actually being parsed, therefore serializing I/O) whereas, I thought execution should be deferred until I try to call collect() or similar on the dataframe. 

My second questions is what the best way is to union the results of these dataframes - is there a way to pass a list of files to read? Ultimately, I really want a single dataframe with all of these results.

```
archives = glob.glob (os.path.join (path, "*.xml"))
dataframes = map (parse, archives)
union = self.sc.parallelize ([])
for d in dataframes:
    union = union.union (d)
```

The parse() function is something like this:
```
            sqlContext = SQLContext (sc)
            dataframe = sqlContext.read.format ('com.databricks.spark.xml'). \
                        options(rowTag='RowTag'). \
                        load (file_uri)
            dataframe = dataframe.select (dataframe["FieldA"], dataframe["FieldB"]). \
                    map (lambda r : translate_row (r))

```

**Comment by stevencox** [comment]
I just learned about SparkContext.union() which seems like a better option for my second question.

**Comment by stevencox** [comment]
@HyukjinKwon, can you help me understand why the code below still reads one file at a time?

1. First, we find a set of XML files and parse them. My intention is to union a set of dataframes without actually executing the Spark pipeline:

```
        archives = glob.glob (os.path.join (self.medline_path, "*.xml"))
        dataframes = map (self.parse_xml, archives)
        union = self.sc.union (dataframes).select (
            dataframe["DateCreated"],
            dataframe["Article"]["Journal"]["JournalIssue"]["PubDate"],
            dataframe["PMID"]). \
            map (lambda row : Medline.map_date_numeric (row))
```
2. The parse_xml function looks like this:
```
def parse_xml (...): ...

@staticmethod
    def parse_medline_xml (data_root, file_name):
        dataframe = None
        ...
        logger.info ("** Parsing {0}".format (file_uri))
        return SQLContext (Medline.sc).read. \
            format ('com.databricks.spark.xml'). \
            options(rowTag='MedlineCitation'). \
            load (file_uri)
```
I had hoped that parse_xml would not actually parse the file so that I could union the resulting dataframes for parallel loading. That's not what happens. There's a long pause as each file is completely parsed.

Thanks for your help.

**Comment by stevencox** [comment]
FYI, for whatever it's worth to others, I 
* Got rid of dataframes
* parallelized the file list as an RDD
* parsed each with ElementTree
* flatMapped the results into a flat RDD
It's easy and perfectly parallel.

**Comment by Bertrandbenj** [comment]
It takes time because if you dont specify a schema, it has to infer it.
though im not a python coder and you seem to have solved your issue

**Comment by anna19954** [comment]
hello, I have met the same problem as you, can you give me an example of how to realize concurrency, I used scala, but I don't know what to do with it
